x-pack/plugins/reporting/server/routes/generate_from_savedobject.ts

```diff
@@ -0,0 +1,157 @@
+ */
+const getJobFromRouteHandler = async (
+  handleRoute: HandlerFunction,
+  handleRouteError: HandlerErrorFunction,
+  request: Request,
+  h: ResponseToolkit,
+  options: { isImmediate: boolean }
+): Promise<QueuedJobPayload> => {
+  const { savedObjectType, savedObjectId } = request.params;
+  let result: QueuedJobPayload;
+  try {
+    const jobParams: JobParams = {
+      savedObjectType,
+      savedObjectId,
+      isImmediate: options.isImmediate,
+    };
+    result = await handleRoute(CSV_FROM_SAVEDOBJECT_JOB_TYPE, jobParams, request, h);
+  } catch (err) {
+    throw handleRouteError(CSV_FROM_SAVEDOBJECT_JOB_TYPE, err);
+  }
+
+  if (get(result, 'source.job') == null) {
+    throw new Error(`The Export handler is expected to return a result with job info! ${result}`);
+  }
+
+  return result;
+};
+
+/*
+ * This function registers API Endpoints for queuing Reporting jobs. The API inputs are:
+ * - "immediate" flag: whether to execute the job up front and make immediate download available
+ * - saved object type and ID
+ * - time range and time zone
+ * - application state:
+ *     - filters
+ *     - query bar
+ *     - local (transient) changes the user made to the saved object
+ */
+export function registerGenerateCsvFromSavedObject(
+  server: KbnServer,
+  handleRoute: HandlerFunction,
+  handleRouteError: HandlerErrorFunction
+) {
+  const getRouteConfig = getRouteConfigFactoryReportingPre(server);
+  const routeOptions = {
+    ...getRouteConfig(() => CSV_FROM_SAVEDOBJECT_JOB_TYPE),
+    validate: {
+      params: Joi.object({
+        savedObjectType: Joi.string().required(),
+        savedObjectId: Joi.string().required(),
+      }).required(),
+      payload: Joi.object({
+        state: Joi.object().default({}),
+        timerange: Joi.object({
+          timezone: Joi.string().default('UTC'),
+          min: Joi.date().required(),
+          max: Joi.date().required(),
+        }).optional(),
+      }),
+    },
+  };
+  const getDocumentPayload = getDocumentPayloadFactory(server);
+
+  // csv: immediate download
+  server.route({
+    path: `${BASE_GENERATE}/immediate/csv/saved-object/{savedObjectType}:{savedObjectId}`,
+    method: 'POST',
+    options: routeOptions,
+    handler: async (request: Request, h: ResponseToolkit) => {
+      /*
+       * 1. Queue a job with getJobFromRouteHandler
+       *   - `isImmediate: true` gets us the complete result data in payload.objects
+       * 2. Copy the completed data stashed in the job as output.content
+       *   - Makes the content available for download
+       * 3. Return a response with CSV content
+       */
+      const queuedJob: QueuedJobPayload = await getJobFromRouteHandler(
+        handleRoute,
+        handleRouteError,
+        request,
+        h,
+        {
+          isImmediate: true,
+        }
+      );
+
+      // FIXME this is REALLY ugly
+      const jobSource: JobDoc = get(queuedJob, 'source.job');
```

Comment:
I talked with @joelgriffith for a bit today about what actually is ugly about this.

1. It's a hack for using the Reporting system of queueing and claiming jobs to generate CSV immediately, and have the API route handler return it in the response immediately
  - This hack stands out, but almost everything else is shared code for immediate and non-immediate
  - Let's maybe double-check that statement
2. A "good" thing about it, is that immediate CSVs create documents in .reporting containing the output, and can be re-downloaded
3. There's mostly bad things about the hack:
  - Mucks around with ESQueue structures; using ESQueue in a way it wasn't meant to be used
  - Not sure if re-download is important, or even if immediate CSV should be featured in the Report listing _at all_

The way to get things to make more sense would be not use `handleRoute` in this handler. Looking back, that's more of a [helper on how to queue up Reporting jobs](https://github.com/elastic/kibana/blob/925998e/x-pack/plugins/reporting/server/routes/index.ts#L23).